src/com/palmergames/bukkit/towny/listeners/TownyPlayerListener.java

```diff
@@ -771,23 +776,80 @@ public void onPlayerChangePlotEvent(PlayerChangePlotEvent event) {
-	 */
+	 * - Handles outlaws entering a town they are outlawed in.
+	 **/
 	@EventHandler(priority = EventPriority.NORMAL)
 	public void onOutlawEnterTown(PlayerEnterTownEvent event) throws NotRegisteredException {
-		
-		Player player = event.getPlayer();		
-		WorldCoord to = event.getTo();
-		Resident resident = TownyUniverse.getInstance().getDataSource().getResident(player.getName());
 
-		if (to.getTownBlock().getTown().hasOutlaw(resident))
-			TownyMessaging.sendMsg(player, Translation.of("msg_you_are_an_outlaw_in_this_town", to.getTownBlock().getTown()));
+		Resident outlaw = TownyUniverse.getInstance().getDataSource().getResident(event.getPlayer().getName());
+		Town town = event.getEnteredtown();
+		TownBlock location = event.getTo().getTownBlock();
+		
+		String[] split = new String[0];
+		long cooldown = TownySettings.getOutlawKickCooldown();
+		
+		// If this resident is an outlaw in this town
+		if (location.getTown().hasOutlaw(outlaw)) {
+			if (!TownySettings.canOutlawsEnterTowns()) {
+				// Are towns warned when there is an outlaw entering?
+				if (TownySettings.doTownsGetWarnedOnOutlaw()) {
+					TownyMessaging.sendPrefixedTownMessage(town, Translation.of("msg_outlaw_town_notify").replace("%s", outlaw.getName()));
+				}
+				// Cooldown messages
+				if (cooldown < 1) {
+					TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kick", town));
+				} else {
+					TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kick_cooldown").replace("%c", TimeMgmt.formatCountdownTime(cooldown)).replace("%s", town.getFormattedName()));
+				}
+				
+				new BukkitRunnable() {
+					@Override
+					public void run() {
+						
+						if (town.hasOutlaw(outlaw)) {
+							
+							Player player = outlaw.getPlayer();
+							
+							// Outlaw has/doesn't have a town.
+							if (!outlaw.hasTown()) {
+								if (player.getBedSpawnLocation() == null) { 
+									player.teleport(town.getWorld().getSpawnLocation());
+								} else {
+									player.teleport(player.getBedLocation());
+								}
+							} else if (outlaw.hasTown()) {
+								try {
+									if (outlaw.getTown().hasSpawn()) {
+										try {
+											SpawnUtil.sendToTownySpawn(player, split, outlaw.getTown(), Translation.of("msg_err_cant_afford_tp"), false, false, SpawnType.TOWN);
+										} catch(TownyException exception) {
+											player.teleport(town.getWorld().getSpawnLocation());
+										}
+									} else {
+										try {
+											SpawnUtil.sendToTownySpawn(player, split, outlaw, Translation.of("msg_err_cant_afford_tp"), false, false, SpawnType.RESIDENT);
+										} catch(TownyException exception) {
+											player.teleport(town.getWorld().getSpawnLocation());
+										}
+									}
+								} catch (NotRegisteredException e) {
+									player.teleport(town.getWorld().getSpawnLocation());
+								}
+								TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kicked", town));
+							}
+						}
```

Comment:
I believe this can be simplified to:
```java
		if (TownyAPI.getInstance().getTown(player.getLocation()) != null && TownyAPI.getInstance().getTown(player.getLocation()) == town && town.hasOutlaw(player.getName())) {
			spawnLoc = town.getWorld().getSpawnLocation();
			if (player.getBedSpawnLocation() != null) 
				spawnLoc = player.getBedSpawnLocation();
			if (outlaw.hasTown() && TownyAPI.getInstance().getTownSpawnLocation(player) != null)
				spawnLoc = TownyAPI.getInstance().getTownSpawnLocation(player);

			TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kicked", town));
			PaperLib.teleportAsync(player, spawnLoc, TeleportCause.PLUGIN);
		}
```

Reply:
Does work, implemented it in the latest commit